Ledger Nano X integration

v-almonacid opened the pull request:
## Tx Signing with ledger using V1 rust bindings (hack needed)
As suggested by Sebastian, I used a hack based on [this PR](https://github.com/Emurgo/yoroi-frontend/pull/160) by Ruslan. In short, a tx is first signed using rust v1 bindings, then it is parsed in JS using `cbor`. The inputs and outputs from rust are provided to Ledger, and finally the witnesses of the original rust tx are replaced by the witnesses given by Ledger. The tx is then re-encoded in JS using `cbor`.
The logic for encoding/decoding rust txs is located in ` src/crypto/byron/util.js`

Note: the `encodeTxAsRust` function doesn't generate *exactly* the same bytes as rust (it differs in a few bytes of metadata I guess). However, the backend seems to decode it correctly.

## Some observed issues
Trying to connect to a device that has been idle for some time seems to fail, and a restart is needed. The best is to do all operations right after inserting the PIN and opening the ADA app (and actually, waiting for at least ~5 seconds after opening the app)

## New dependencies
- `@cardano-foundation/ledgerjs-hw-app-cardano` 1.0.8: Ledger operations
- `@ledgerhq/react-native-hw-transport-ble` 5.13.0: Ledger connection
- `react-native-ble-plx` 1.1.1: bluetooth connection. **An older version of this library was installed** because latest version (2.x) uses RN 0.60+. We'll need to upgrade this library once RN version is upgraded too.
- `cbor` 5.0.2: Needed for hack explained above
- `jsc-android`: Required by `@ledgerhq/react-native-hw-transport-ble`

## Features for follow up PR
- Address verification (UI needed)
- USB-cable connection support (this can only be supported for Android though)
- perhaps some minor UI improvements
- The "learn more about using yoroi with ledger" must point to some relevant link (video or tutorial), which currently doesn't exist now.

## Modified screens

### Add wallet screen

![image](https://user-images.githubusercontent.com/7271744/80956183-fc2a5b80-8e00-11ea-98dd-eb6f48d10303.png)

### Send screen

Only added a few instructions, but might require some improvements.

![image](https://user-images.githubusercontent.com/7271744/80956195-03ea0000-8e01-11ea-9aa1-6f1aceae4e7c.png)

## New screens

### Check Nano X screen

![image](https://user-images.githubusercontent.com/7271744/80956206-0c423b00-8e01-11ea-8b41-a882676ae418.png)

#### Scrolling down shows more text in Android

![image](https://user-images.githubusercontent.com/7271744/80956220-149a7600-8e01-11ea-89ef-6d8c846309ce.png)

### Connect Nano X screen

![image](https://user-images.githubusercontent.com/7271744/80956672-ef5a3780-8e01-11ea-831f-119e3da9875a.png)

![image](https://user-images.githubusercontent.com/7271744/80956681-f41eeb80-8e01-11ea-9f9d-e2dd1294afa6.png)

![image](https://user-images.githubusercontent.com/7271744/80956685-f84b0900-8e01-11ea-8fb0-dcbd5f3ef6de.png)

### Save Nano X screen 

![image](https://user-images.githubusercontent.com/7271744/80956691-faad6300-8e01-11ea-9815-3e7c92d3b155.png)

v-almonacid committed:
add ledger wallet creation screens and stories

v-almonacid committed:
add dependencies and configurations for bluetooth and ledger support

v-almonacid committed:
Implement bluetooth connection. Add progress step bar and more UI changes

v-almonacid committed:
implement wallet creation with ledger

v-almonacid committed:
add support for tx signing with ledger

SebastienGllmt commented on ios/dummy_ble.swift:
```diff
@@ -0,0 +1,13 @@
+//  NOTE: after upgrading to RN 0.60+ this file can be safely deleted
+//
+//  Created by Vicente Almonacid on 08/04/2020.
+//  Copyright © 2020 Facebook. All rights reserved.
```
I think this copyright is incorrect

v-almonacid replied:
That was automatically generated by Xcode (I only added the note). What Copyright should be used?

SebastienGllmt commented on ios/emurgo-Bridging-Header.h:
```diff
@@ -0,0 +1,4 @@
+//
+//  Use this file to import your target's public headers that you would like to expose to Swift.
```
Do we still need these two header files?

v-almonacid replied:
yep

SebastienGllmt commented on package.json:
```diff
@@ -26,6 +26,8 @@
     "@babel/core": "7.4.5",
     "@babel/node": "7.0.0",
     "@babel/runtime": "7.4.5",
+    "@cardano-foundation/ledgerjs-hw-app-cardano": "^1.0.8",
```
Are we okay with non-strict dependencies for all of these?

SebastienGllmt commented on src/__mocks__/config.js:
```diff
@@ -10,29 +10,37 @@ import env from '../env'
+    NETWORK: 'Testnet', // for now, assume Shelley testnet by default
     PROTOCOL_MAGIC: 764824073,
     API_ROOT: 'https://shelley-itn-yoroi-backend.yoroiwallet.com/api',
+    SEIZA_STAKING_SIMPLE: (ADA: string) =>
```
I'm confused why this file was modified in this PR. Additionally, why are these part of the mock but removed from the config.js in the same PR?

v-almonacid replied:
This is not really relevant. I just updated the mock config. I did a small change that broke some test, fixed it and changed a few extra things.

SebastienGllmt commented on src/components/Send/ConfirmScreen.js:
```diff
@@ -78,6 +81,32 @@ const handleOnConfirm = async (
     }
   }
 
+  // TODO: add a wrapper for setSendingTransaction()
```
multiple TODOs

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/CheckNanoXScreen.js:
```diff
@@ -0,0 +1,104 @@
+  },
+})
+
+// TODO
```
TODO

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/ConnectNanoXScreen.js:
```diff
@@ -0,0 +1,288 @@
+
+type Props = {
+  intl: intlShape,
+  defaultDevices: ?Array<Device>, // for UI design prototyping
```
?

v-almonacid replied:
This is used by storybook.

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/ConnectNanoXScreen.js:
```diff
@@ -0,0 +1,288 @@
+    TransportBLE.observeState({
+      next: (e) => {
+        if (this.bluetoothEnabled == null && !e.available) {
+          // consider adding a specific string since this is a common error
```
?

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/ConnectNanoXScreen.js:
```diff
@@ -0,0 +1,288 @@
+    this.startScan()
+  }
+
+  // componentDidUpdate = (prevProps) => {
```
can we remove this?

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/ConnectNanoXScreen.js:
```diff
@@ -0,0 +1,288 @@
+    })
+  }
+
+  // TODO
```
TODO

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/ConnectNanoXScreen.js:
```diff
@@ -0,0 +1,288 @@
+    this.startScan()
+  }
+
+  // TODO
```
TODO

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/DeviceItem.js:
```diff
@@ -0,0 +1,71 @@
+})
+
+type Props = {
+  onSelect: (device: any) => any, // TODO
```
add types and maybe make `Props` exact

SebastienGllmt commented on src/crypto/byron/ledgerUtils.js:
```diff
@@ -0,0 +1,331 @@
+export const createLedgerSignTxPayload = async (
+  unsignedTx: PreparedTransactionData,
+  txsBodiesMap: TxBodiesResponse,
+  // addressedUtxos: Array<AddressedUtxo>,
```
Presumably you will need this to know the derivation path for the key of each input

SebastienGllmt commented on src/crypto/byron/ledgerUtils.js:
```diff
@@ -0,0 +1,331 @@
+): Array<InputTypeUTxO> {
+  return inputs.map((input) => {
+    // TODO
+    // verifyAccountLevel(addressingInfo)
```
I'm not sure you can check this in Yoroi Mobile since we don't have anything other than accounts

SebastienGllmt commented on src/crypto/byron/ledgerUtils.js:
```diff
@@ -0,0 +1,331 @@
+    address58: txOutput.address,
+    amountStr: txOutput.value,
+  }))
+  // TODO: verify if order is important
```
Just needs to match the same order what the Ledger signs. Protocol itself doesn't care about order of outputs

SebastienGllmt commented on src/crypto/byron/ledgerUtils.test.js:
```diff
@@ -0,0 +1,99 @@
+jestSetup.setup()
+// jest.setTimeout(30 * 1000)
+
+describe('prepare HW tx', () => {
```
I created some Jest tests in my latest merged PR in the Yoroi extension that you might be interested in

SebastienGllmt commented on src/crypto/wallet.js:
```diff
@@ -266,71 +305,124 @@ export class Wallet {
+      }
+    } catch (e) {
+      Logger.error('wallet::_integrityCheck', e)
+      // TODO: open general error dialog. Use intl
```
TODO

v-almonacid committed:
remaining work:
- add instructions for tx sending
- add InconsisteState wallet exception and associated intl messages
- moved some strings to global-messages
- disabled confirm button while handling tx confirmation through ledger
- some missing types
- completed other misc TODOs

v-almonacid committed:
fix linking of react-native-ble-plx on iOS

v-almonacid committed:
handle HW connection error

SebastienGllmt commented:
> Trying to connect to a device that has been idle for some time seems to fail, and a restart is needed. The best is to do all operations right after inserting the PIN and opening the ADA app (and actually, waiting for at least ~5 seconds after opening the app)

In the browser extension, I think we had to increase the timeout, but there is a browser limit so after the timeout we automatically fail the prompt. This is different from what you're doing (different transport mode and not a browser), but just so you're aware.

SebastienGllmt commented on package.json:
```diff
@@ -59,6 +63,7 @@
     "react-markdown": "^4.0.6",
     "react-native": "0.59.10",
     "react-native-background-timer": "^2.1.1",
+    "react-native-ble-plx": "^1.1.1",
```
Are you okay with loose version number?

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/ConnectNanoXScreen.js:
```diff
@@ -0,0 +1,257 @@
+  navigation: Navigation,
+}
+
+type State = {
```
prefer exact type

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/SaveNanoXScreen.js:
```diff
@@ -0,0 +1,138 @@
+  )
+}
+
+type ExternalProps = {
```
prefer exact types

SebastienGllmt commented on src/crypto/byron/ledgerUtils.js:
```diff
@@ -0,0 +1,351 @@
+import type BluetoothTransport from '@ledgerhq/react-native-hw-transport-ble'
+
+// these are defined in LedgerConnectStore.js in yoroi-frontend
+type LedgerConnectionResponse = {
```
prefer strict for all of these

SebastienGllmt commented on src/crypto/byron/ledgerUtils.js:
```diff
@@ -0,0 +1,351 @@
+
+// Hardware wallet device Features object
+// borrowed from HWConnectStoreTypes.js in yoroi-frontend
+export type HWFeatures = {
```
I removed a lot of these from the Yoroi Extension. Notably, I moved it down to 

```
export type HWFeatures = {|
  Vendor: string,
  Model: string,
  DeviceId: string,
|};
```
but I also use Label for the default name.
However, not that almost all of these are Trezor only (I think `vendor` is the only one provided by Ledger)

v-almonacid replied:
Agree. I'll simplify. But why removing the app version? I think it's important if we need to debug any issue eventually (even though there should be very few version upgrades, I guess)

SebastienGllmt replied:
The app version doesn't really make sense to store because it can change at any time. Somebody can start Yoroi, then while the app is still running, do an update of their Ledger device. If you ever need to do something based off the version, you would need to query it at the time

v-almonacid committed:
improve error handling plus other misc changes

v-almonacid committed:
move to react-native-webview

v-almonacid committed:
more ui/ux improvements

v-almonacid committed:
add NSBluetoothAlwaysUsageDescription for nightly scheme

v-almonacid committed:
fix linking of BleClient for nightly and staging schemes

v-almonacid committed:
bump react-native-device-info to 5.5.7 (no UIWebView from v3)

SebastienGllmt commented on src/crypto/byron/ledgerUtils.js:
```diff
@@ -0,0 +1,414 @@
+  return false
+}
+
+// note: e.statusCode === ErrorCodes.ERR_CLA_NOT_SUPPORTED probably means
```
> probably

Have we confirmed this, or are we leaving it as "probably" for now?

v-almonacid replied:
Yes, "probably" is not precisely what I meant. Theoretically that error rises when the wrong app is opened in ledger, but I've observed it a few times when no app is opened at all and a tx is requested. So I said "probably" because the error is "probably" due to the user not having opened the ADA app instead of having opened the wrong app.

SebastienGllmt commented on src/components/WalletInit/ConnectNanoX/CheckNanoXScreen.js:
```diff
@@ -0,0 +1,103 @@
+})
+
+// TODO
+const url = '' // 'https://yoroi-wallet.com/...'
```
Why is both this and `ledgerMessages.internetConnected` commented out?

v-almonacid replied:
@SebastienGllmt `url` should be set once we have a video ready, we said we will add it once the video is ready.

About the `ledgerMessages.internetConnected` string, would you agree if we just remove it? I think I initially added it because we had it in the extension, but in the extension you really need internet to communicate with the http bridge. In mobile you don't strictly need internet.

SebastienGllmt commented on src/i18n/global-messages.js:
```diff
@@ -79,6 +79,60 @@ export const txLabels = defineMessages({
+  // connection requisites
+  enableTransport: {
+    id: 'global.ledgerMessages.enableTransport',
+    defaultMessage: '!!!Enable bluetooth and location services (Android only).',
```
Do we want to specify "Android only" in the string instead of only displaying this on Android devices?

SebastienGllmt commented:
Feedback:
> remove wallet no longer relevant message asking you to make sure you write down your 15 words.
> "change password" no longer relevant with Ledger confirmation for this kind of wallet
> easy transaction confirmation" no longer relevant with Ledger confirmation for this kind of wallet

v-almonacid committed:
replace react-native-qrcode by react-native-qrcode-svg

v-almonacid committed:
remove UI components not relevant for HW wallets
